### PR TITLE
118 redirect docs

### PR DESCRIPTION
**Pull request by tatiana-nspcc**
Redirect from `/` into `v1/docs/`.

Close https://github.com/nspcc-dev/neofs-rest-gw/issues/118.

### CODE REVIEW

File: handlers/api.go [hunk 1]
```diff
@@ -197,13 +203,35 @@ func (a *API) docMiddleware(handler http.Handler) http.Handler {
 		if strings.HasPrefix(r.URL.Path, docsPrefix) {
 			fh.ServeHTTP(w, r)
 		} else if r.URL.Path == "" || r.URL.Path == "/" {
-			http.Redirect(w, r, "/v1/docs/", http.StatusFound)
+			// Get basePath from spec
+			basePath, err := getBasePath("spec/rest.yaml")
+			if err != nil {
+				a.log.Info("cannot get basePath from spec/rest.yaml", zap.Error(err))
+				handler.ServeHTTP(w, r)
```
Comment (smallhive) [inline handlers/api.go:210]: Do we need `return` here? Right now we serve request, and later, redirect to `/docs`

File: handlers/api.go [hunk 2]
```diff
@@ -196,12 +202,36 @@ func (a *API) docMiddleware(handler http.Handler) http.Handler {
...
+			}
+			http.Redirect(w, r, basePath+"/docs/", http.StatusFound)
 		} else {
 			handler.ServeHTTP(w, r)
 		}
 	})
 }
 
+func getBasePath(path string) (string, error) {
```
Comment (roman-khimov) [inline handlers/api.go:219]: Doing this for every request to `/` doesn't seem to be fun. And this path is likely present somewhere anyway. Can we avoid this?

File: handlers/api.go [hunk 3]
```diff
@@ -196,12 +202,36 @@ func (a *API) docMiddleware(handler http.Handler) http.Handler {
...
+	// Read the YAML file
+	yamlFile, err := os.ReadFile(path)
```
Comment (roman-khimov) [inline handlers/api.go:221]: Oh, and you're requiring this file to be present on the server. It won't be. There is `gen/restapi` package with JSON though.

File: handlers/api.go (hunk 2 above)
Comment (roman-khimov) [reply handlers/api.go:219]: I'd expect it to be somewhere around
```
        swaggerSpec, err := loads.Analyzed(restapi.SwaggerJSON, "")
```

File: handlers/api.go [hunk 4]
```diff
@@ -197,13 +207,34 @@ func (a *API) docMiddleware(handler http.Handler) http.Handler {
...
+	if swaggerSpec.loaded {
+		return swaggerSpec.basePath, nil
+	}
+
+	sp, err := loads.Analyzed(restapi.SwaggerJSON, "")
+	if err != nil {
+		return "", err
+	}
+
+	swaggerSpec.basePath = sp.BasePath()
```
Comment (roman-khimov) [inline handlers/api.go:233]: But isn't it racy? You can have several requests doing this at the same time.

File: handlers/api.go (hunk 1 above)
Comment (tatiana-nspcc) [reply handlers/api.go:214]: Yep, you are right

Comment (tatiana-nspcc) [reply handlers/api.go:214]: Yep, you are right

File: handlers/api.go [hunk 5]
```diff
@@ -3,7 +3,9 @@ package handlers
 import (
 	"context"
 	"fmt"
+	"gopkg.in/yaml.v3"
```
[the remark this answers is not in the document]
Comment (tatiana-nspcc) [reply handlers/api.go]: Fixed.

File: handlers/api.go (hunk 4 above)
Comment (tatiana-nspcc) [reply handlers/api.go:241]: Yep, you're probably right. I've added mutex in the `spec` struct.

Comment (tatiana-nspcc) [reply handlers/api.go:241]: Yep, you're probably right. I've added mutex in the `spec` struct.

Comment (roman-khimov) [reply handlers/api.go:241]: Why don't you do this in `init()` then? No bool, no mutex, a single execution that does the job and then you always have the result of it.

Comment (tatiana-nspcc) [reply handlers/api.go]: I've moved it in init() function.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**